internal/llrp/behavior.go

```diff
@@ -0,0 +1,993 @@
+	pwrMinToMax []TransmitPowerLevelTableEntry
+	freqInfo    FrequencyInformation
+
+	// report is the collection of information we want expect a Reader to report.
+	// LLRP has a data compression "feature" that allows Readers to omit some parameters
+	// if the value hasn't changed "since the last time it was sent".
+	report TagReportContentSelector
+	// lastData is the value of tag parameter the last time it was reported.
+	lastData TagReportData
+
+	nGPIs, nFreqs uint16
+	nSpecsPerRO   uint32
+	allowsHop     bool
+	stateAware    bool
+}
+
+// ImpinjDevice embeds BasicDevice to provide some Impinj-specific Behavior implementations.
+//
+// Impinj isn't compliant with the following elements of the LLRP standard:
+// - UHF Modes incorrectly report BLF (in Hz) instead of BDR (in bps).
+// - UHF Modes include "Autoset" modes with IDs > 1000,
+//   for which the parameter values are made up;
+//   the actual mode used is one of the non-Autoset modes,
+//   but the Reader interprets the given ModeID as a hint
+//   for it to choose which of those it thinks is best.
+// - Truncate actions during Select (i.e., C1G2Filter.T) are not supported and must be 0.
+// - Per-antenna configurations are not compliant in general,
+//   but you can set transmit power and receive sensitivity per-antenna.
+//
+// Additionally, the following limitations have been observed:
+// - If Hopping is True, we've observed only a single HopTable.
+// - Depending on the firmware and specific Reader type, 2-5 C1G2 Filters are available.
+// - There seem to be 5 real UHF Modes, though which are available
+//   is limited by Reader version and Firmware.
+//   All of them use Multi-Reader or Dense-Reader spectral masks,
+//   a DR of 64/3, and do not permit Tari selection (min==max):
+//   - Mode0 is their fastest, at 640kbps using Tari 6.25us, PIER 1.5, and FM0.
+//   - Mode1 they call "Hybrid"; it's the same as Mode0,
+//     but uses Miller2 backscatter encoding, so the BDR is 320kbps.
+//   - Mode2 is for "Dense" environments and has a BDR of 68.5kbps.
+//   - Mode3 is for even denser environments, with the slowest BDR at 21.25kbps.
+//     They call it MaxMiller because it uses Miller8 backscatter encoding.
+//   - They don't support State Aware Filtering, at least not directly.
+//     There is a custom parameter for "Search Mode" which essentially does it.
+type ImpinjDevice struct {
+	BasicDevice
+}
+
+func NewBasicDevice(c *GetReaderCapabilitiesResponse) (*BasicDevice, error) {
+	if c == nil || c.LLRPCapabilities == nil || c.GeneralDeviceCapabilities == nil ||
+		c.RegulatoryCapabilities == nil || c.C1G2LLRPCapabilities == nil {
+		return nil, errMissingCapInfo("capabilities")
+	}
+
+	regCap := c.RegulatoryCapabilities
+	if regCap == nil || regCap.UHFBandCapabilities == nil ||
+		len(regCap.UHFBandCapabilities.TransmitPowerLevels) == 0 {
+		return nil, errMissingCapInfo("power levels",
+			"RegulatoryCapabilities", "UHFBandCapabilities", "TransmitPowerLevels")
```

Comment:
`RegulatoryCapabilities` will never be missing at this point. Already covered by line 146

Reply:
This is just calling a helper function that constructs an `error` with a helpful message that includes the path within the struct that failed validation, so these string parts are here for that reason. The resulting error message is something like `missing LLRP power levels from RegulatoryCapabilities.UHFBandCapabilities.TransmitPowerLevels`. Specifying them separately allows a single "constructor"(ish) for this error type regardless of whether a "path" applies, and as a bonus, is a bit cleaner to write the code (since the path parts can be knocked down).

Reply:
OK